File: cses/leetcode/cpp/75.cpp
// use three pointers, and then follow algo just like quicksort
#include <bits/stdc++.h>
using namespace std;
class Solution
{
public:
    void sortColors(vector<int> &nums)
    {
        int n = nums.size();
        int i = 0, j = 0, k = n - 1;
        while (j < k)
        {
            if (nums[j] == 0)
            {
                swap(nums[i], nums[j]);
                        }
            else if (nums[j] == 2)
            {
                swap(nums[j], nums[k]);
                k--;
            }
            j++;
        }
        return;
    }
};
int main()
{
}

File: leetcode/go/75.go
package main

import "fmt"

func sortColors(nums []int) {
	n := len(nums)
	for i, j, k := -1, n, 0; k < j; {
		if nums[k] == 0 {
			i++
			nums[i], nums[k] = nums[k], nums[i]
			k++
		} else if nums[k] == 2 {
			j--
			nums[j], nums[k] = nums[k], nums[j]
		} else if nums[k] == 1 {
			k++
		}
	}
	fmt.Println(nums)
}
